extensions/amp-date-countdown/0.1/amp-date-countdown.js

```diff
@@ -0,0 +1,283 @@
+  'th': ['ปี', 'เดือน', 'วัน', 'ชั่วโมง', 'นาที', 'วินาที'],
+  'tr': ['yıl', 'ay', 'gün', 'saat', 'dakika', 'saniye'],
+  'vi': ['năm', 'tháng', 'ngày', 'giờ', 'phút', 'giây'],
+};
+
+export class AmpDateCountdown extends AMP.BaseElement {
+
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @const {function(!Element)} */
+    this.boundRendered_ = this.rendered_.bind(this);
+
+    //Note: One of end-date, timestamp-ms, timestamp-seconds is required.
+    /** @private {string} */
+    this.endDate_ = this.element.getAttribute('end-date');
+
+    /** @private {number} */
+    this.timestampMs_ = Number(this.element.getAttribute('timestamp-ms'));
+
+    /** @private {number} */
+    this.timestampSeconds_ = Number(this.element.getAttribute('timestamp-seconds'));
+
+    /** @private {number} */
+    this.offsetSeconds_ = Number(this.element.getAttribute('offset-seconds')) || DEFAULT_OFFSET_SECONDS;
+
+    /** @private {string} */
+    this.locale_ = (this.element.getAttribute('locale')) || DEFAULT_LOCALE;
+
+    /** @private {string} */
+    this.whenEnded_ = (this.element.getAttribute('when-ended')) || DEFAULT_WHEN_ENDED;
+
+    /** @private {string} */
+    this.biggestUnit_ = (this.element.getAttribute('biggest-unit')) || DEFAULT_BIGGEST_UNIT;
+
+    /** @private {!Object} */
+    this.localeWordList_ = this.getLocaleWord_(this.locale_);
+
+    /** @const {!../../../src/service/template-impl.Templates} */
+    this.templates_ = Services.templatesFor(this.win);
+  }
+
+  /** @override */
+  buildCallback() {
+    const epoch = this.getEpoch_() + (this.offsetSeconds_ * 1000);
+    this.tickCountDown_(new Date(epoch) - new Date());
+  }
+
+  /** @override */
+  layoutCallback() {
+    const epoch = this.getEpoch_() + (this.offsetSeconds_ * 1000);
+    let differentBetween = new Date(epoch) - new Date() - 1000; //substract 1000 here because of buildCallback show the initial time
+    const delay = 1000;
+    return Promise.resolve(this.countDownTimer_ = this.win.setInterval(() => {
+      this.tickCountDown_(differentBetween);
+      differentBetween -= delay;
+    }, delay));
+  }
+
+  /** @override */
+  unlayoutCallback() {
+    this.win.clearInterval(this.countDownTimer_);
+    return true;
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /**
+   * @param {!Object} items
+   * @return {!Promise}
+   * @private
+   */
+  renderItems_(items) {
+    return this.templates_.findAndRenderTemplate(this.element, items)
+        .then(this.boundRendered_);
+  }
+
+  /**
+   * @param {number} differentBetween
+   * @private
+   */
+  tickCountDown_(differentBetween) {
+    const data = this.getYDHMSFromMs_(differentBetween);
+    if (this.whenEnded_ === 'stop' && differentBetween < 1000) {
+      Services.actionServiceForDoc(this.element)
+          .trigger(this.element, 'timeout', null, ActionTrust.HIGH);
```

Comment:
This event can not be `HIGH-TRUST` since it is not tied to a user-interaction and AMP does not allow non-user-initiated events to run arbitrary actions and cause page layout changes. 

Low-trust would be fine. Keep in mind low-trust event can only trigger certain other low-trust actions such as `amp-animation` and `amp-video` actions. Playing with `visibility` on `amp-animation` hopefully should handle many cases where things need to become shown/hidden as result of the `timeout` event (with visibility, they can't cause content shift, so UX is fine and it is done at compositor layer so perf is fine too).

Reply:
wait a minute, if i am using ActionTrust.LOW, the current `timeout` action will encounter error as figure below.

![image](https://user-images.githubusercontent.com/4065175/37263599-d57b1b50-25e3-11e8-8810-ae441ed86461.png)


``` html
<h1 id="sample">
    When Timer hits 0, will hide the timer itself and hide this message.
</h1>
<h1 id="sample2" hidden>
    When Timer hits 0, will hide the timer itself and display this message.
</h1>
<amp-date-countdown id="ampdate" end-date="2018-03-12T10:55:00+08:00" on="timeout: ampdate.hide, sample.hide, sample2.show;" height="235" width="500" when-ended="stop" locale='en'>
```

Reply:
Anyway, I have changed from `ActionTrust.LOW` back to `ActionTrust.HIGH`

Reply:
@aghassemi

Reply:
@leonalicious Unfortunately we can not allow events that are not explicitly triggered by a user action to be HIGH trust. HIGH trust events can cause major relayout, submission of form and other destructive changes to an AMP page which are now allowed in AMP without user interactions. 

This means you can not use `hide` and `show` actions as result of timeout ( they cause relayout and content to shift). However you can use `amp-animation` to change `visibility` of an item to `hidden` or `visible`. This is allowed because it does not cause relayout. 

So instead of hide you can do:
```
<amp-animation id="hideTransition"
  layout="nodisplay">
  <script type="application/json">
    [{
      "duration": "1",
      "fill": "both",
      "selector": ".#ItemThatNeedsToHidden",
      "keyframes": [{
          "visibility": "hidden"
        }
      ]
    }]
  </script>
</amp-animation>


<amp-date-countdown on="timeout:hideTransition.play">...
```

Reply:
changed actionTrust from HIGH to LOW and use animation to set the visibility although i personally think there are better approach on this.

Reply:
Thanks.